### ISSUE TITLE
Tracking issue for potential adoption of JSpecify nullness annotations

### ISSUE THREAD

**Issue by kevinb9n** [jOOQ/jOOQ#12949]
### Background

* jOOQ is satisfied with its adoption of the JetBrains nullness annotations, and this isn't a particular source of pain.
* JetBrains and Google and several other organizations involved in nullness analysis are building a "standard" artifact by consensus, with full generics support and well-specified semantics, under the name jspecify.org.

("Standard" in quotes because bringing it into OpenJDK _per se_ is a very long-term background thread. They see our project as the only viable route toward that goal. We still believe users will see it as the "one clear choice" we've been lacking -- so long as we don't mess it up.)

To learn about JSpecify: I'm working on a more ideal overview suited to your perspective, but in the meantime you might pick up on a few differences from JetBrains annotations by scanning our current [early-draft user guide](https://jspecify.dev/user-guide.html).

### What is this issue

jOOQ's _motivation_ to migrate to the new annotations when they are ready might be anywhere on a continuum from "yawn" to "how do we help" (all fair).

If you're okay with it, this issue could be a useful place to track:

* What specific obstacles to migration can we think of?
    * examples: it needs to have feature X; it needs to be in state Y before we'd even do a test migration; it has deal-breaker issue Z; the change itself would be disruptive because...
* What might make migration more appealing / motivating / worth spending time on?

This discussion could progress toward your eventual decision TO adopt _or_ that you will NOT adopt. In the meantime, the discussion would be immensely valuable to us as we make design and prioritization decisions.

### A note on timing

An actual test-adoption would be a bit premature (we have this mostly done for Guava, and can show that, but then, we are hacking on our own tools as we do it!).

But as things ramp up then the various pressures to start setting design decisions "in stone" will as well, and we'd hope to get your complaints about them before that happens.

But we have to make that worth your time to actually do. This issue is just a first step to open an ongoing conversation about all this between us.

(previous: https://github.com/jOOQ/jOOQ/issues/6244#issuecomment-1020939730)

**Comment by lukaseder** [comment]
Thanks again for reaching out. I would be thrilled to replace the JetBrains annotations currently in use by something more "standard", meaning something that experiences even more adoption by the community.

### These are must haves:

- JetBrains replaces their own annotations by the JSpecify ones as a sign of commitment.
- IntelliJ supports the annotations out of the box in their introspections for various languages
- kotlin and scala support the annotations in their type systems
- The JSpecify ones support Java 8 in some way (multiple jars, multi-release jars, etc.) as the jOOQ Professional and Enterprise Editions will support Java 8 for quite another while.
- While JSpecify mostly seems to target Java and kotlin development, I'd like it to also embrace Scala's nullability: https://docs.scala-lang.org/scala3/reference/other-new-features/explicit-nulls.html

The must-have-ness stems from the point that I don't want to have two sets of nullability annotations on the jOOQ API in parallel. So, while testing them in parallel can be fine, a release will migrate completely or not at all.

### These are big pros, but not strictly must haves:

- Oracle / Open JDK publicly confirms their endorsement of the JSpecify endeavour even if they don't (yet) commit to tackling this topic in the JDK itself.

### These, I have to evaluate first:

- jOOQ currently works with the "kotlin interop model", distinguishing between
  - kotlin `T!`, Java `T`: Unspecified nullability
  - kotlin `T`, Java `@NotNull T`: Guaranteed not-nullness
  - kotlin `T?`, Java `@Nullable T`: Guaranteed nullness

  From what I understand, JSpecify does not (yet) want to add a `@NotNull` annotation, but instead allow for certain scopes (module, package, class) to be `@NullMarked`. My intuition is that this won't work for "pragmatic" libraries like jOOQ, which migrate gradually to the new world. For example, jOOQ so far added annotations on:

  - Return types (only)

  jOOQ does not yet annotate

  - Type variables (which is quite hard to get right and a lot of work, given the weirdness of Java's collection API with respect to nulls)
  - Parameter types (which is a lot of work, and not strictly necessary for kotlin / scala users)

  The relevant task to annotate more API is here: https://github.com/jOOQ/jOOQ/issues/10780

  So, from jOOQ's perspective, it would be *very* desirable to continue to gradually migrate towards annotating the API (keep in mind that jOOQ's API is exceptionally huge, given its internal DSL nature, so this is much more work than for other libraries), and the current `@NullMarked` scoping doesn't suffice.

  In addition to that, jOOQ takes a special stance on nullability of scalar SQL values. Much like in Java, in SQL, any value can be `NULL`. Nullability is not an attribute of a column expression's *type*, but rather a dynamic constraint on it, which is modified by much more than just scalar expression operators. For example, if you specify:

  ```sql
  create table t (i int not null);
  ```

  Then you might be tempted to assert `T.I` is of type `Field<@NotNull Integer>`, because that would be so convenient for projections of `T.I`. But that assumption is incorrect once you left join `T`, or create other more complex expressions involving `T`:

  ```sql
  -- Anything outer joined can be NULL again
  select t.i from x left join t on p

  -- Empty scalar subqueries always produce NULL
  select (select t.i from t where false)

  -- Who on earth thinks of this case?
  select t.i from t where false group by grouping sets ((t), ())
  ```

  All the kotlin fans keep complaining about this issue here https://github.com/jOOQ/jOOQ/issues/10212, thinking that the above Would-Never-Happen-To-Them™, but I really don't like making unsound promises in a flawed type system. A defensive solution is making everything `Field<T?>` (but that's terribly inconvenient). A pragmatic solution is to keep using the Java interoperability of kotlin and making everything `Field<T!>` (but they don't like Java code in their otherwise kotlin codebase). The wrong solution is to promise `Field<T>` and then run into NPEs in any of the above edge cases. I don't think I can ever model SQL's nullability in any client language with an *internal* DSL, it's only possible via external DSLs (e.g. using jOOQ's parser and generating code from parsed SQL).

  Long story short, while maybe I can be talked into re-thinking all of this, I *think* I need kotlin's three-valued nullability system: `T!`, `T`, and `T?` (kotlin) or `T`, `@NotNull T`, `@Nullable T` (Java). I personally also think that allowing `T!` types in kotlin would have been a good choice in kotlin itself, though this question of mine didn't get much love yet: https://stackoverflow.com/q/61814732/521799, so maybe, jOOQ's use-case is exceptional. Though I still think that from an adoption perspective, you should have `@NotNull` as well, as an escape hatch.

**Comment by lukaseder** [comment]
Related discussion on why kotlin won't have a denotable `T!` type: https://github.com/Kotlin/KEEP/issues/82

**Comment by kevinb9n** [comment]
This is outstanding feedback, better than expected.

Hopefully having this discussion a bit here in this issue is not too disorganized/annoying for now; we'll be making sure we have the individual issues filed we need.


> ### These are must haves:
> * JetBrains replaces their own annotations by the JSpecify ones as a sign of commitment.
> * IntelliJ supports the annotations out of the box in their introspections for various languages
> * kotlin and scala support the annotations in their type systems
> * Oracle / Open JDK publicly confirms their endorsement of the JSpecify endeavour even if they don't (yet) commit to tackling this topic in the JDK itself.

Okay, for now I'll check what sort of brief public statements they'd be comfortable giving us. Perhaps of the form "assuming X and Y are achieved then our intention is to do Z..." Hopefully this'd give you some idea of what to expect.


> * The JSpecify ones support Java 8 in some way (multiple jars, multi-release jars, etc.) as the jOOQ Professional and Enterprise Editions will support Java 8 for quite another while.

Yes, 8 is a requirement. Currently trying the mr-jar thing.


> * While JSpecify mostly seems to target Java and kotlin development, I'd like it to also embrace Scala's nullability: https://docs.scala-lang.org/scala3/reference/other-new-features/explicit-nulls.html

Ondřej Lhoták is a member but not active, so this is a good reminder to us, thanks.


>   From what I understand, JSpecify does not (yet) want to add a `@NotNull` annotation,

I think it's more like this: once we discovered that there's a fully "sane" picture of the world without it, and next convinced ourselves that adding it in later wouldn't seem that detabiilizing, then we just threw it on the back burner. I don't think we feel strongly about which way it will go for 1.0 yet. We do have a giant document somewhere with about ten pros and ten cons of adding it to pore over (but truthfully, quite a few of the arguments are subtle as hell).

But if it facilitates adoption, that will be compelling.

The issue is https://github.com/jspecify/jspecify/issues/4 but I can't recommend trying to read all that.


> jOOQ so far added annotations on:
>   
>   * Return types (only)
>   
>   jOOQ does not yet annotate
>   
>   * Type variables (which is quite hard to get right and a lot of work, given the weirdness of Java's collection API with respect to nulls)
>   * Parameter types (which is a lot of work, and not strictly necessary for kotlin / scala users)

Useful. I hadn't thought about adoption happening in this direction. I'm guessing you aren't running nullness analysis on your own code, then? If true, do you have _some_ kind of support for ensuring you're adding the annotations correctly? If not, have you wanted it? Curious about this kind of thing.


>   In addition to that, jOOQ takes a special stance on nullability of scalar SQL values.

Aha. Unless I misunderstand, I think this sounds like Yet Another Case of a pattern we've seen many times:

* A certain type usage does technically (at least rarely) include null
* Yet, in many very common cases, we know it absolutely won't
* Now no one can ever be happy

Some are SO lopsided that we can round up and classify them non-nullable (`System.out`!). So the really vexing cases are the ones that don't quite meet that bar... and there are plenty.

Our analogue of https://github.com/Kotlin/KEEP/issues/82 is more or less https://github.com/jspecify/jspecify/issues/71, but I do have more to add / clean up there.

We still hope we will have a better solution than `@UnspecifiedNullness`.

**Comment by lukaseder** [comment]
Glad to hear this is useful!

> Perhaps of the form "assuming X and Y are achieved then our intention is to do Z..."

Yes, that would be helpful

> The issue is [jspecify/jspecify#4](https://github.com/jspecify/jspecify/issues/4) but I can't recommend trying to read all that.

I'll be happy to read through it. I haven't given all the pros and cons proper thought, perhaps reading this will help change my mind.

> Useful. I hadn't thought about adoption happening in this direction. I'm guessing you aren't running nullness analysis on your own code, then?

Not the usual one, no. I turned on Eclipse's analysis experimentally, but it's currently full of false positives. Unsurprisingly, as only the public API was annotated, (almost) none of the internal APIs. Though, so far, NPEs haven't been too big of an issue with jOOQ's development.

> If true, do you have _some_ kind of support for ensuring you're adding the annotations correctly?

There are some ASM based API checks, which validate some API rules. In addition to that, there's instrumentation that is active during integration tests to verify if `@NotNull` annotated API ever returns `null`, as well as some parts of the other API, which does return `null` is annotated `@Nullable`. Given the high test coverage jOOQ has, this covers most of the current API.

This works so far because the annotations are `ElementType.METHOD` annotations, and can be discovered from their `CLASS` retention. Not sure how such an approach would work with `TYPE_USE` annotations, which are erased.

> If not, have you wanted it?

Yes, I'd definitely want better validation, though again, I'm not sure yet what that could look like.

> We still hope we will have a better solution than `@UnspecifiedNullness`.

I understand that this isn't the desired solution for many. Though, in jOOQ's case, it would be quite useful and enable usage of the `@NullMarked` annotation, keeping the `@NotNull` noise at a minimum.

**Comment by kevinb9n** [comment]
> > The issue is [jspecify/jspecify#4](https://github.com/jspecify/jspecify/issues/4) but I can't recommend trying to read all that.
> 
> I'll be happy to read through it. I haven't given all the pros and cons proper thought, perhaps reading this will help change my mind.

To be clear, I'm expressing that I think it will actively leave you more confused than before. It's the same with any issue that's gotten that long or dates back that many (sigh) years. There has been such a changing vocabulary and set of assumptions that it will make no sense. We need a better fresh summary. 

> Not sure how such an approach would work with `TYPE_USE` annotations, which are erased.

This sounds approximately like a known javac bug we're trying to fix, but I don't have the details in my head atm.
But, I also think that we will probably need to have a solution for anyone who needs to stay on declaration annotations for whatever reason.

> > We still hope we will have a better solution than `@UnspecifiedNullness`.
> 
> I understand that this isn't the desired solution for many. Though, in jOOQ's case, it would be quite useful and enable usage of the `@NullMarked` annotation, keeping the `@NotNull` noise at a minimum.

The "hope" I mean is that we'll have a solution that's equally satisfying (or failing that, nearly enough so). We will try.

**Comment by cpovirk** [comment]
As one of kevinb9n's teammates, I also thank you for the comments!

A few things I wanted to ask you about:

- The runtime instrumentation (for unexpected nulls) is something that we and others expect to want someday. It sounds straightforward but painful to implement for someone not already deeply familiar with bytecode manipulation. I assume that there's not already some preexisting tool that anyone can plug in to do this, but I'd like to ask just in case there happens to be :)
- For type-use annotations, the most _fundamental_ problem we're aware of is [JDK-8226216](https://bugs.openjdk.java.net/browse/JDK-8226216), which prevents annotation processors (and similar plugins) from seeing type-use annotations in code read from .class files. (This is the one Kevin mentioned that we have work underway to fix.) The annotations are still visible when reading from source files. And the data is _in_ the class files, so Java's reflection APIs will ([mostly](https://bugs.openjdk.java.net/browse/JDK-8202469)) expose it, and tools like ASM can, too. Still, consuming type-use annotations can still require a lot of extra work, and I imagine that tools like the instrumenter discussed above would need to be updated. Anyway, I said I had a question here, and it is: Do you know if there are other cases in which the data is outright dropped (like javac does in the class-file case), or it the data available but requires extra effort to consume?
- If you are annotating only return types, then which tools benefit from having `@NonNull` there? Inside Google, our nullness checking comes mostly from Kotlin and the Checker Framework (and opportunistically in IntelliJ). As I understand it, adding `@NonNull` to a _return type_ is mostly a no-op from their perspective. Is this different for Scala or other tools? Or do you need `@NonNull` on return types for the tools I'm familiar with, too? (I know there are cases in which it can matter, so maybe jOOQ happens to hit those more often than our average project does.)

Also, if you already opened https://github.com/jspecify/jspecify/issues/4, I suggest closing it in favor of https://github.com/jspecify/jspecify/issues/229 and the _additional_ issues that Kevin will be writing :)

**Comment by cpovirk** [comment]
Actually, now I see that I have overlooked the obvious connection between two of my questions:

One reason to annotate returns types with `@NonNull` is so that runtime instrumentation can check that such methods never return `null` :)

Now, theoretically, we could do that with a `@NonNull` annotation that isn't recognized by _any_ tools _except_ the tool that performs the instrumentation. But I grant that is hardly the most natural way to approach this.

Anyway, let me know if there's more to it than that. Thanks again for your time.

**Comment by lukaseder** [comment]
> @kevinb9n But, I also think that we will probably need to have a solution for anyone who needs to stay on declaration annotations for whatever reason.

The workaround that I've thought about in the past would be for an annotation processor to produce a side car file (XML or whatever), which can then be processed by tests much more easily than if anything needed to be done in the annotation processor directly. That by itself could be an interesting project. An XML format to represent the byte code as seen by an annotation processor!

> @kevinb9n The "hope" I mean is that we'll have a solution that's equally satisfying (or failing that, nearly enough so). We will try.

Well, an 80/20 solution can still do the trick, it worked well for kotlin. It might not work for *some* APIs, like jOOQ...

> @cpovirk I assume that there's not already some preexisting tool that anyone can plug in to do this, but I'd like to ask just in case there happens to be :)

I'm currently using Javassist to insert Java code before and after all methods of interest to me, e.g.

```java
method.insertAfter("if ($_ == null) throw new NullPointerException(\"NotNull method returns null: " + method.getLongName() + "\");");
```

Nothing beautiful, but it does the trick. Today, I might use byte-buddy, but for the simple byte code manipulation I'm doing, Javassist was sufficient.

> @cpovirk Do you know if there are other cases in which the data is outright dropped (like javac does in the class-file case), or it the data available but requires extra effort to consume?

I haven't done much more research. To me, improving nullability behaviour is mostly a favour for the 5% (10%?) kotlin/jOOQ users. The bulk of Java/jOOQ users, which also includes the bulk of the paying customers, hasn't been too keen in requesting improvements in this area, so I haven't invested too much time into this topic (yet).

I'm pretty sure I had discussed things with @amaembo from JetBrains in the past, on twitter (e.g. https://twitter.com/tagir_valeev/status/1449044592075083782, but I had a different discussion in mind, which I can't find now). Perhaps he had a specific case in mind?

> @cpovirk If you are annotating only return types, then which tools benefit from having `@NonNull` there?

So far, mostly, the kotlin compiler and the IDE. `T` types generate less noise than `T!` types, especially when you can write `Result<Record>>` rather than `Result<Record!>!` and much deeper nesting.

> @cpovirk Is this different for Scala or other tools?

I haven't investigated Scala 3 yet. Scala 2 doesn't use union types for nulls, yet, so it doesn't apply.

> @cpovirk  (I know there are cases in which it can matter, so maybe jOOQ happens to hit those more often than our average project does.)

The most interesting case for jOOQ is the one jOOQ cannot handle so far, i.e. the `Field<T>` type, which helps decide whether `T` or `T?` is being projected by a query.

> @cpovirk One reason to annotate returns types with `@NonNull` is so that runtime instrumentation can check that such methods never return `null` :)

Yes, though, so far, I'm profiting from the fact that the Jetbrains annotation is both `TYPE_USE` and `METHOD` (for backwards compatibility reasons?), so I'm taking the much easier method annotation analysis route...

> @cpovirk Now, theoretically, we could do that with a `@NonNull` annotation that isn't recognized by _any_ tools _except_ the tool that performs the instrumentation. But I grant that is hardly the most natural way to approach this.

To be sure: I'm impartial to using an actual `@NonNull` annotation. A `@NullMarked` scope annotation where everything is non-null by default is fine, too, as long as there's also something like `@NullUnspecified` for those cases where nullness cannot be specified in jOOQ.

Currently, this can be achieved using the JetBrains annotations by annotating `@NotNull` and `@Nullable` explicitly, and leaving everything unspecified as it is. `@NullMarked` would obviously be far less noisy.

**Comment by kevinb9n** [comment]
(minor tweak: you will have to write `<T extends @Nullable Object>`, which looks super sad, but we can give you a list of like 6 reasons it's better.)

I think you'll have what you want here. Indeed, parametric nullness is distinct from unspecified nullness. (In fact, these will become different even just in Java itself, as part of specialized generics / Valhalla).

So in JSpecify: usages of type variables are unaffected by the presence of `@NullMarked` on an enclosing scope. While every _non-type-variable_ type usage has a nullness operator from the "real three", a type variable (except a non-null-bounded one) is a different case. A 4th kind of nullness makes the whole thing sound (4/3)^3 as complicated, but it is really just the natural extension of type variable behavior into the nullness axis: you're allowed to do something with it if and only if it preserves the type variable's ability to be safely substituted later with *any* type in bounds.

So it behaves as you expect, and in fact this is _sometimes_ a workaround for a lack of `@PolyNull`, as `cast()` could be null-passthrough on `Class<@Nullable String>` but not on `Class<String>`. (Not a good workaround, as `Class`'s `T` has no business being nullable at all.

HTH

**Comment by lukaseder** [comment]
> (minor tweak: you will have to write `<T extends @Nullable Object>`, which looks super sad, but we can give you a list of like 6 reasons it's better.)

Makes sense, of course, thanks

**Comment by tamaro-skaljic** [comment]
[Heise Developer](https://www.heise.de/blog/Null-Problemo-Bessere-Null-Checks-in-Java-mit-JSpecify-9820606.html) has just informed me that [JSpecify has released version 1.0.0](https://jspecify.dev/blog/release-1.0.0/).

**Comment by lobaorn-bitso** [comment]
Hi @lukaseder , given that Spring 7 and Spring Boot 4 began paving the way for broader adoption of JSpecify (https://spring.io/blog/2025/03/10/null-safety-in-spring-apps-with-jspecify-and-null-away) and recently JUnit (https://github.com/junit-team/junit5/pull/4557) also began fully using it, have you re-evaluated using JSpecify annotations in jOOQ since?

**Comment by lukaseder** [comment]
@lobaorn-bitso Is there any obvious benefit right now over the status quo, given:

1. The status quo works
2. The change is a change (and thus work in terms of changing, communicating, etc. etc.)
3. We're also using other Jetbrains annotations, which we still want to use after a hypothetical change

I don't see any benefit, unless the library goes into the JDK.
